8304006: jlink should create the jimage file in the native endian for the target platform

jaikiran opened the pull request:
Can I please get a review for this change which proposes to fix the issue reported in https://bugs.openjdk.org/browse/JDK-8206890?

The `jlink` command allows a `--endian` option to specify the byte order in the generated image. Before this change, when such a image was being launched, the code would assume the byte order in the image to be the native order of the host where the image is being launched. That would result in failure to launch java, as noted in the linked issue.

The commit in this PR, changes relevant places to not assume native order and instead determine the byte order by reading the magic bytes in the image file's header content.

A new jtreg test has been added which reproduces the issue and verifies the fix.

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8304006](https://bugs.openjdk.org/browse/JDK-8304006): jlink should create the jimage file in the native endian for the target platform (**Enhancement** - P4)


### Reviewers
 * [Alan Bateman](https://openjdk.org/census#alanb) (@AlanBateman - **Reviewer**)
 * [Mandy Chung](https://openjdk.org/census#mchung) (@mlchung - **Reviewer**) ⚠️ Review applies to [258a62d7](https://git.openjdk.org/jdk/pull/11943/files/258a62d73338d1bfc57d7a3b66dc4ce97936e7c3)
 * [Roger Riggs](https://openjdk.org/census#rriggs) (@RogerRiggs - **Reviewer**) ⚠️ Review applies to [258a62d7](https://git.openjdk.org/jdk/pull/11943/files/258a62d73338d1bfc57d7a3b66dc4ce97936e7c3)


### Contributors
 * Mandy Chung `<mchung@openjdk.org>`

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/11943/head:pull/11943` \
`$ git checkout pull/11943`

Update a local copy of the PR: \
`$ git checkout pull/11943` \
`$ git pull https://git.openjdk.org/jdk.git pull/11943/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 11943`

View PR using the GUI difftool: \
`$ git pr show -t 11943`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/11943.diff">https://git.openjdk.org/jdk/pull/11943.diff</a>

</details>


### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/11943#issuecomment-1378845543)

AlanBateman commented on src/java.base/share/classes/jdk/internal/jimage/ImageHeader.java:
```diff
@@ -180,4 +185,39 @@ int getStringsOffset() {
+     * @return The ByteOrder if one could be determined. Else null.
+     * @throws IOException If any error occurred reading the contents of the image file.
+     */
+    static ByteOrder tryDetectByteOrder(final Path imageFile) throws IOException {
```
Style-wise I think I would try to keep it consistent with the existing code if you can, meaning the noise/over-use of finals can mostly be removed if you can.

jaikiran replied:
Done. I've updated the PR to remove all the `final` usage from this part of the code and made it consistent with the current code.

AlanBateman commented:
I hope @JimLaskey or @sundararajana can review this.

jaikiran commented:
While running tier testing, one existing test showed that I hadn't taken into account SecurityManager checks in the new code that I had added. I have now fixed that part and pushed an update to the PR.

There's one additional unrelated test failing (`TestMaxCachedBufferSize`) in a odd way and I'll investigate it separately.

AlanBateman commented:
I'm in two minds on whether this should be "fixed" in libjimage. I'm wondering if we libjimage should assume native order and instead explore having jlink determine the endianness at link time for the target platform - this might be looking at some property in the packaged module for java.base (java.base.jmod). @JimLaskey Do you have any opinion on this?

For the image reader used by jrtfs then I think it has to work with both as the file system could be created with java.home set in the env to a run-time image for a complete different platform.

AlanBateman commented:
@jaikiran I've been mulling over this a bit further. When jlink is cross linking, as in creating a run-time image for a different target platform, then I think we should strive to have it create the jimage file in the native endian for the target platform. Right now this means the user needs to help out by providing the --endian option. For example, if the user is on linux-x64 but generating a run-time image for solaris-sparcv9 or aix-ppc, then it requires specifying jlink --endian=big to ensure that the jimage file is in the endianness needed at runtime. I don't disagree that having the 2-arg ImageFileReader::open retry with the opposite of the endian parameter will work but I don't think this is the right thing to do. Instead I think we need to consider determine the endian from the target java.base module. Right now the ModuleTarget class file attribute is not enough but a short term fix might be for jlink to have a simple mapping of known ModuleTarget values to endianness. Overall, it's probably low priority as the most likely users of this will be environments building for embedded platforms.

JimLaskey commented on src/java.base/share/classes/jdk/internal/jimage/ImageHeader.java:
```diff
@@ -180,4 +187,48 @@ int getStringsOffset() {
+        // try with the other byte order
+        ByteOrder altByteOrder = bb.order() == ByteOrder.BIG_ENDIAN
+                ? ByteOrder.LITTLE_ENDIAN : ByteOrder.BIG_ENDIAN;
+        bb.flip();
```
Don't you just want to rewind here?

jaikiran replied:
Hello Jim, you are right `rewind()` would be a better API to use here.

JimLaskey commented on src/java.base/share/classes/jdk/internal/jimage/ImageHeader.java:
```diff
@@ -180,4 +187,48 @@ int getStringsOffset() {
+        ByteBuffer bb = ByteBuffer.wrap(bytes);
+        if (bb.getInt() == MAGIC) {
+            return bb.order();
+        }
```
I would have just added `public static final int REVERSE_MAGIC = 0xDADAFECA;` and add an else clause. to compare to REVERSE_MAGIC. Much simpler.

JimLaskey commented on src/java.base/share/native/libjimage/imageFile.cpp:
```diff
@@ -281,8 +281,20 @@ ImageFileReader* ImageFileReader::open(const char* name, bool big_endian) {
         // Failed to open.
-        delete reader;
-        return NULL;
+        if (reader != NULL && reader->_invalid_magic) {
```
Similar here. Flip the byte order if matches IMAGE_REVERSE_MAGIC.

JimLaskey commented:
I'm in agreement with Alan that this change seems suspicious. That is, you should not attempt to run with non-native byte order.  That said, I can see where the java reader could read non-native as an analysis tool. So,  C code should continue to be an error if not the native byte order. The java code, might be useful for tools but not sure.

mlchung commented:
> When jlink is cross linking, as in creating a run-time image for a different target platform, then I think we should strive to have it create the jimage file in the native endian for the target platform. .... Instead I think we need to consider determine the endian from the target java.base module.

Having jlink to determine the endianness for the target platform is a good idea.  This will improve its usability.  Currently jlink looks at the `ModuleTarget` attribute to determine the target platform, which can be extended to include endianness.

> Right now the ModuleTarget class file attribute is not enough but a short term fix might be for jlink to have a simple mapping of known ModuleTarget values to endianness.

This is a viable short-term solution.

jaikiran commented:
Thank you everyone for the review and the suggestions.

> Right now the ModuleTarget class file attribute is not enough but a short term fix might be for jlink to have a simple mapping of known ModuleTarget values to endianness. Overall, it's probably low priority as the most likely users of this will be environments building for embedded platforms.

I'll experiment with the short term fix. For now, I'll put this PR in draft and will reopen when it's ready for review.

jaikiran commented:
> Right now the ModuleTarget class file attribute is not enough but a short term fix might be for jlink to have a simple mapping of known ModuleTarget values to endianness. Overall, it's probably low priority as the most likely users of this will be environments building for embedded platforms.

I've now updated this PR to revert the previous commits and instead follow this suggestion to infer the endianness of the target platform while generating the image.

The updated PR now continues to allow users to provide `--endian` option while generating an image through `jlink`. If any value is provided for that option, the image generation will continue to honour that value. However, if no value is provided for that option, then during image generation, we will determine the endianness of the target platform, by using the (OpenJDK specific) `ModuleTarget` attribute of target `java.base` module's `module-info` class. This determination is best-effort and if the value cannot be determined, then we default to the current platform's endianness (like we have been doing so far).

This commit hardcodes the mapping between a `ModuleTarget`'s architecture and endianness and this mapping is borrowed from what we already do in the OpenJDK build file https://github.com/openjdk/jdk/blob/master/make/autoconf/platform.m4#L26

No new tests have been added in this change, because I couldn't think of a way to  test the cross platform `jlink` execution. Existing tests in `test/jdk/tools/jlink` continue to pass with this change. I've triggered our CI runs to run the tiered tests to make sure nothing regresses.

The commit in this PR doesn't deprecate the --endian option and I think we can do that as a separate issue.

jaikiran commented:
I've updated the PR title to now point to the new https://bugs.openjdk.org/browse/JDK-8304006

AlanBateman commented on src/jdk.jlink/share/classes/jdk/tools/jlink/internal/JlinkTask.java:
```diff
@@ -884,6 +904,31 @@ private static String findModuleName(Path modInfoPath) {
+                        "ppc64", "s390",
+                        "s390x", "sh",
+                        "sparc", "sparcv9" -> ByteOrder.BIG_ENDIAN;
+                default -> ByteOrder.nativeOrder();
```
When cross linking and the target platform is unknown then it might be better to have this fail rather than defaulting to the endianness of the current platform, otherwise this risks create a run-time image that is not usable.

I see you've listed a bunch of unknown platforms in the list. If getNativeEndianOfTargetPlatform fails when it encounters an unknown platform then help downstream porters to know that there is something to look at here.

jaikiran replied:
> When cross linking and the target platform is unknown then it might be better to have this fail rather than defaulting to the endianness of the current platform, otherwise this risks create a run-time image that is not usable.

I wasn't sure if we wanted to fail such cases when generating the image, but like you say, the generate image would be unusable anyway. I've now followed your suggestion to fail in such cases.

> I see you've listed a bunch of unknown platforms in the list. If getNativeEndianOfTargetPlatform fails when it encounters an unknown platform then help downstream porters to know that there is something to look at here.

I borrowed that list from our current build file here https://github.com/openjdk/jdk/blob/master/make/autoconf/platform.m4#L26. I think it's likely that it isn't the current supported list of platforms. I checked the `porters-dev` mailing list and I see that recently we have created this wiki which lists the supported platforms https://wiki.openjdk.org/display/HotSpot/Ports. Based on that I've trimmed down this list in the code (I continue to include `loongarch64` because I recently saw some PR which dealt with that architecture in mainline).  Having said that, it might be that this too isn't the fully accurate list, so please let me know if this needs additional changes.

AlanBateman reviewed (commented):
Thanks for dropping the changes to the jimage code and changing this to choose the endianness when cross linking.

jaikiran commented on src/jdk.jlink/share/classes/jdk/tools/jlink/internal/JlinkTask.java:
```diff
@@ -812,7 +813,26 @@ private static class ImageHelper implements ImageProvider {
+                    ModuleReference ref = javaBase.get().reference();
+                    if (ref instanceof ModuleReferenceImpl modRefImpl
+                            && modRefImpl.moduleTarget() != null) {
+                        targetPlatform = modRefImpl.moduleTarget().targetPlatform();
```
Hello Alan,

>  If getDefaultModulePath() is the parent of the location containing java.base then it can use the current endianness.

I've now updated this PR to implement this suggestion. The `ModuleTarget` attribute in java.base's `module-info.class` will only be checked if the java.base isn't from the current execution environment's default jmods location.

jaikiran replied:
> Also just to say that if it would be an error if java.base is not in the Configuration, same thing if there is no packaged module for java.base.

I've now added `assert`s to expect this module to be present.

jaikiran commented on src/jdk.jlink/share/classes/jdk/tools/jlink/internal/JlinkTask.java:
```diff
@@ -884,6 +904,30 @@ private static String findModuleName(Path modInfoPath) {
+                return ByteOrder.nativeOrder();
+            }
+            String archName = targetPlatform.substring(index + 1);
+            return switch (archName) {
```
Hello Jim, yes it feels odd to be listing it in the code here. Adding it as a runtime resource (like a properties file?) too seems a bit odd to me. Perhaps this should be written out during build to some templated Java file which can then be used in this code? Something like a `SupportedPlatforms.java.template` file under the `jdk.tools.jlink.internal` package which has a:
``` java
ByteOrder getByteOrder(String platform)
```
method which gets generated during  the build?

Either way, I think there would then need to be a build file of officially supported platforms which can be sourced into that template file. I'm not too familiar with the porting process, is there such a precendence for it? Do porters update such files to add to the list as a supported platform?

AlanBateman commented on src/jdk.jlink/share/classes/jdk/tools/jlink/internal/JlinkTask.java:
```diff
@@ -831,6 +879,28 @@ private static class ImageHelper implements ImageProvider {
+                        // the "java.base" module file used to create the image
+                        (PrivilegedExceptionAction<Boolean>) () -> Files.isSameFile(javaBasePath,
+                                currentPlatformJmods.resolve(Path.of("java.base.jmod"))));
+            } catch (PrivilegedActionException e) {
```
jlink doesn't run with a security manager so no need for the AccessController.doPriv.

Assuming the name of the jmod file is probably okay.

jaikiran replied:
> jlink doesn't run with a security manager so no need for the AccessController.doPriv.

Thank you for that detail - I've now removed the SecurityManager checks from this newly introduced code.

> Assuming the name of the jmod file is probably okay.

Were you thinking of exploded builds? I'm not too familiar with exploded builds, but after you commented here about the `.jmod` extension, I checked the existing code in `JLinkTask` and I can see that we seem to allow the module to be packaged as a `.jar` too, like here https://github.com/openjdk/jdk/blob/master/src/jdk.jlink/share/classes/jdk/tools/jlink/internal/JlinkTask.java#L834. Do you think this cross platform image creation would be relevant/used with exploded builds and if so, should I update the PR to address it?

jaikiran replied:
> Do you think this cross platform image creation would be relevant/used with exploded builds and if so, should I update the PR to address it?

Actually, now that I think about it, this `isJavaBaseFromCurrentPlatform()` method will get executed even for regular (same platform) image generation. Which means if the `java.base` module file doesn't end in a `.jmod` extension then this part of the code can run into `java.nio.file.NoSuchFileException` . I will generate an exploded build and see if anything needs to be done.

jaikiran replied:
I decided to remove the use of hardcoded `.jmod` extension while checking the paths. I have updated the PR to use a slightly different technique which doesn't require using hardcode file names for determining if this a cross-platform image generation.

Manual testing shows that this change continues to work fine when identifying cross-platform image generation. Existing automated tests too pass fine for same platform image generation.

jaikiran commented on src/jdk.jlink/share/classes/jdk/tools/jlink/internal/JlinkTask.java:
```diff
@@ -884,6 +954,27 @@ private static String findModuleName(Path modInfoPath) {
+                        "sparc", "sparcv9" -> ByteOrder.BIG_ENDIAN;
+                default -> null;
+            };
+        }
```
> To add to Jim's comment, the other thing we think about here is mapping the value of the ModuleTarget attribute to endianness, meaning don't parse it to extract the architecture.

So that would then mean updating the build file where we create these jmod file(s) to use a value for `--target-platform` that even includes the endianness in the string value, example: `aix-ppc64-big` and `macos-aarch64-little`. So a change here https://github.com/openjdk/jdk/blob/master/make/CreateJmods.gmk#L252. Did I understand it correctly?

AlanBateman commented on src/jdk.jlink/share/classes/jdk/tools/jlink/internal/JlinkTask.java:
```diff
@@ -831,6 +876,19 @@ private static class ImageHelper implements ImageProvider {
+            // check if the current platform's "jmods" directory is the parent of
+            // the "java.base" module file used to create the image
+            return Files.isSameFile(javaBasePath,
+                    currentPlatformJmods.resolve(Path.of("java.base.jmod")));
```
You can reduce this to resolve("java.base.jmod").

mlchung commented on src/jdk.jlink/share/classes/jdk/tools/jlink/internal/JlinkTask.java:
```diff
@@ -884,6 +947,27 @@ private static String findModuleName(Path modInfoPath) {
+        // returns the endianness of the target platform, if the target platform is known
+        // and supported for creating an image through jlink. Else returns null.
+        private static ByteOrder getNativeEndianOfTargetPlatform(String targetPlatform) {
+            int index = targetPlatform.indexOf("-"); // of the form <operating system>-<arch>
```
`Platform::parsePlatform` is the utility method to parse `ModuleTarget`.   It can be updated to include additional architectures.

AlanBateman replied:
> `Platform::parsePlatform` is the utility method to parse `ModuleTarget`. It can be updated to include additional architectures.

Alternatively, don't parse it. If we go with Jim's suggestion of a resource file then it is just a simple mapping of the ModuleTarget value, e.g the entry for macox-x64 would be:
```
macos-amd64.endianness = little
```

jaikiran replied:
Hello Mandy, Alan, Jim,

I've updated this PR to take into account these suggestions. I went ahead with what Mandy suggested and enhanced the existing (internal) `jdk.tools.jlink.internal.Platform` `record`to additional parse these architectures. It was anyway doing it for a select few.

In addition to that, I also moved the architecture to endianness mapping into this `Platform` class itself. I followed Jim's suggestion to move the mapping out of the code. The `Platform` code now reads this from a `endianness.properties` file which is an internal resource of the `jdk.jlink` module. However, I'm having second thoughts about this part. I'm guessing that when Jim suggested moving this to a resource, it was probably because the code resided within the `JLinkTask` class. Now that I've moved this to an existing `Platform` class which is solely meant for things like these, I feel that we could probably just hard code these architecture to endianness mapping within this class itself, instead of reading it from a properties file. I'm looking for some inputs on what you all think would be the best way to proceed.

Alan, as for the `osname-arch=little/big` vs `arch=little/big` mapping style, I decided to use the `arch=little/big` and let the Platform class parse the architecture out of the string in the `ModuleTarget`. I did it for a couple of reasons:

1. The `Platform` class already has the necessary logic to do that, plus it has to do that anyway (for us to implement the suggestion that Mandy noted about using this in `DefaultImageBuilder`)

2. As far as I could see, the OS name shouldn't play any role in the endianness, so leaving that out felt easier to deal with since we wouldn't have to think of what OS name, arch combinations would be valid.

If you or others feel that it's better to stick with the `osname-arch=little/big` approach, instead of this one, please do let me know and I'll go ahead and do that change.

AlanBateman replied:
>  If you or others feel that it's better to stick with the osname-arch=little/big approach, instead of this one, please do let me know and I'll go ahead and do that change.that out felt easier to deal with since we wouldn't have to think of what OS name, arch combinations would be valid.

My preference would be something like target.properties so we have one place with the properties for target platform when cross linking. ${ModuleTarget}.${property} could work as keys. At some point we will need to extend or replace the ModuleTarget as it too basic to describe the target platform, that's why I was hoping to avoid parsing it here.

jlink cannot be reliably used to cross link to target a different JDK release. Reasons include changes to the run-time image, jimage changes, and plugins that are deeply tied to java.base or other modules. I mention this because JDK 11 was the last release to build solaris-sparcv9, so I think you can this and several others from the properties file. Probably just stick the the builds that are possible in the main line.

jaikiran replied:
Thank you Alan for these additional inputs. I've now updated the PR to use a `target.properties` to map the `osname-arch` to a endianness. The implementation in `Platform` class will no longer use the arch substring to decide the endianness and instead will use the complete `osname-arch` for determining the endianness.

Additionally, I have trimmed down the number of `osname-arch` mapping to what I think is the current supported ones in mainline. If this list is inaccurate (which is possible), then please do let me know.

mlchung commented on src/jdk.jlink/share/classes/jdk/tools/jlink/internal/JlinkTask.java:
```diff
@@ -811,8 +813,51 @@ private static class ImageHelper implements ImageProvider {
+                    // this is an attempt to build a cross-platform image. We now attempt to
+                    // find the target platform's arch and thus its endianness from the java.base
+                    // module's ModuleTarget attribute
+                    Optional<ResolvedModule> javaBase = cf.findModule("java.base");
```
`ModuleTarget` is read by `DefaultImageReader::storeFiles` as well.   Perhaps it should be refactored so that the target `Platform` can be passed to `DefaultImageBuilder`.

jaikiran replied:
I've updated the PR to pass along this already parsed target platform to the `DefaultImageBuilder`. This code is new to me and I think I made sure (by checking references and the package names) that I only updated the internal classes without impacting any user facing APIs. If however, this impacts any user facing API, please do let me know.

jaikiran commented:
Existing tests in `test/jdk/tools/jlink` and `test/jdk/tools/jimage` continue to pass with these changes and a manual test of building a cross platform image shows that it picks up the correct endianness of the target platform (and fails with an error if the target platform's endianness isn't known)

mlchung commented on src/jdk.jlink/share/classes/jdk/tools/jlink/internal/JlinkTask.java:
```diff
@@ -806,13 +809,59 @@ private static class ImageHelper implements ImageProvider {
+                    if (this.order == null) {
+                        // unsupported target platform
+                        throw new IOException(
+                                taskHelper.getMessage("err.unsupported.target.platform",
```
IIUC, the current jlink implementation does not fail if the target platform is unknown as long as `--endian` option is specified.   So I think `--endian` may be useful to keep (rather than taking it out).   This error key should be `err.unknown.target.endianness` and the message can suggest to use `--endian` option to specify.

jaikiran replied:
I've updated the key to match what you suggest here.

mlchung commented on src/jdk.jlink/share/classes/jdk/tools/jlink/resources/jlink.properties:
```diff
@@ -148,6 +148,8 @@ err.cannot.read.module.info=cannot read module descriptor from {0}
 err.not.modular.format=selected module {0} ({1}) not in jmod or modular JAR format
 err.signing=signed modular JAR {0} is currently not supported,\
 \ use --ignore-signing-information to suppress error
+err.cannot.determine.target.platform=cannot determine target platform for image generation
```
```suggestion
err.cannot.determine.target.platform=cannot determine target platform from {0}
```

`{0}` can be the path to `java.base.jmod`  (i.e. `ModuleReference::location`)

jaikiran replied:
Done. Additionally, if the `ModuleReference::location` is unknown (it's an `Optional`) then we will print `cannot determine target platform from java.base module`

mlchung commented on src/jdk.jlink/share/classes/jdk/tools/jlink/resources/jlink.properties:
```diff
@@ -148,6 +148,8 @@ err.cannot.read.module.info=cannot read module descriptor from {0}
 err.signing=signed modular JAR {0} is currently not supported,\
 \ use --ignore-signing-information to suppress error
+err.cannot.determine.target.platform=cannot determine target platform for image generation
+err.unsupported.target.platform=image generation for target platform {0} is not supported
```
```suggestion
err.unsupported.target.endianness=Unknown native byte order for target platform {0}`
```

`{0}` is `targetPlatformVal`

jaikiran replied:
Done

jaikiran commented on src/jdk.jlink/share/classes/jdk/tools/jlink/internal/JlinkTask.java:
```diff
@@ -540,7 +542,7 @@ private static Path toPathLocation(ResolvedModule m) {
+    private static ImageHelper createImageProvider(JlinkConfiguration config,
                                                      Path retainModulesPath,
                                                      boolean ignoreSigning,
                                                      boolean bindService,
```
Fixed

jaikiran commented on src/jdk.jlink/share/classes/jdk/tools/jlink/internal/JlinkTask.java:
```diff
@@ -806,13 +809,59 @@ private static class ImageHelper implements ImageProvider {
+                    boolean verbose,
+                    PrintWriter log) throws IOException {
+            if (order != null) {
+                this.targetPlatform = Platform.runtime();
```
That's a good catch. I've now fixed it as you suggested.

jaikiran commented on src/jdk.jlink/share/classes/jdk/tools/jlink/internal/Platform.java:
```diff
@@ -128,6 +180,16 @@ private static Architecture toArch(String archName) {
+            case "s390"            -> Architecture.s390;
+            case "s390x"           -> Architecture.s390x;
+            case "sparc"           -> Architecture.SPARC;
+            case "sparcv9"         -> Architecture.SPARCv9;
```
I've now updated the PR to trim down the target platforms to match what we have in `target.properties`. Additionally, `is64Bit` has been enhanced too.

I would like to note that, in an intermediate version of this PR, I made `is64Bit` return true for `arm` architecture. I did so because from the materials I found, ARM supports both 32 bit and 64 bit architectures, but the recent versions are 64 bit. So I thought it would be appropriate to consider it 64 bit.

However, I then looked up the `platform.m4` build file and it marks `arm` architecture as 32 bit https://github.com/openjdk/jdk/blob/master/make/autoconf/platform.m4#L57. So I've updated the PR to return false from `is64Bit` from `arm` arch.

jaikiran commented:
Thank you Mandy for the review, I've updated the PR with the changes that you suggested. Existing jtreg tests in `test/jdk/tools/jlink` `test/jdk/tools/jimage` continue to pass and a manual test shows that for cross platform jlink, the code picks up the endianness correctly and also throws an error if an explicit endianness doesn't match the java.base module's target platform.

mlchung commented on src/jdk.jlink/share/classes/jdk/tools/jlink/internal/JlinkTask.java:
```diff
@@ -806,13 +809,57 @@ private static class ImageHelper implements ImageProvider {
+                    PrintWriter log) throws IOException {
+            if (order != null) {
+                this.order = order;
+                String targetPlatformVal = readJavaBaseTargetPlatform(cf);
```
This should also do the fast-path check `isJavaBaseFromCurrentPlatform(javaBasePath)` as in line 838-860.  I suggest to refactor line 838-860 to return the target platform (either runtime platform or from java.base in the given module path).   

I also suggest to include `endianness` in the `Platform` record class.

I sent you a patch offline with these suggestions.

jaikiran replied:
Thank you Mandy for the review and the improved patch. I've applied it and tests continue to pass. I've updated the PR with these changes.

jaikiran commented:
/contributor add @mlchung

AlanBateman commented on src/jdk.jlink/share/classes/jdk/tools/jlink/internal/Platform.java:
```diff
@@ -24,12 +24,30 @@
+        } catch (IOException e) {
+            throw new ExceptionInInitializerError(e);
+        }
+        KNOWN_ENDIANNESS = p;
```
The naming/usage suggests the value is solely for endianness but I think we want target properties to be able to support other attributes of the target too. This means "endianness" in the property name or the property value is something like a comma separate list of attributes, I don't have a strong opinion on this. But maybe your proposal is just to get the target.properties in place and we change it later?

mlchung replied:
I read this as for endianness only and so the filename should be "target-endianness.properties" or something like that.   OTOH,  I like keeping it `target.properties` and add `.endianness` as the property name suffix will make it extensible to support other attributes easily.

jaikiran replied:
Given these inputs, I've now updated the PR to use `.endianness` as a suffix to individual keys in the properties file, to allow us reuse this `target.properties` for other details.

Existing tests continue to pass and manual testing shows the correct endianness is detected during cross platform image creation.

I'll trigger a tier testing to make sure nothing breaks.

mlchung commented on src/jdk.jlink/share/classes/jdk/tools/jlink/builder/DefaultImageBuilder.java:
```diff
@@ -144,16 +143,34 @@ public Platform getTargetPlatform() {
+     * @param targetPlatform target platform of the image
+     * @throws IOException
+     * @throws NullPointerException If any of the params is null
+     * @since 21
```
you can drop this as this is not a public API.

jaikiran replied:
Removed in latest update of the PR.

mlchung commented on src/jdk.jlink/share/classes/jdk/tools/jlink/internal/Platform.java:
```diff
@@ -24,12 +24,42 @@
+                String platform = key.substring(0, key.indexOf(ENDIANNESS_KEY_SUFFIX));
+                if (platform.isEmpty()) {
+                    throw new InternalError("Incorrect key '" + key + "'");
+                }
```
If you scan all entries, this can probably check the value is valid (`little` or `big`) and store them in a `Map<String, ByteOrder>` so that `parsePlatform` can simply get the byte order by `KNOWN_ENDIANNESS.get(platformString)`.

Or the other option can just be doing in the `parsePlatform` method and not needed to scan all entries.

```
   String v = KNOWN_ENDIANNESS.getProperty(platformString + ENDIANNESS_KEY_SUFFIX);
```

jaikiran replied:
Hello Mandy, I was in two minds when I added that iteration logic. I wasn't sure if that was necessary or we could just check for the values when looking up. I have now updated the PR to remove the iteration and just rely on looking up the key, as you suggested, and then checking the value.

jaikiran commented:
> I'll trigger a tier testing to make sure nothing breaks.

tier tests revealed that we have a `test/jdk/tools/jlink/plugins/CDSPluginTest.java` test which tries to simulate a cross platform jlink image generation. Given our changes in this PR, which only reads the `ModuleTarget` of java.base when the java.base module comes from a different file than the one in the current platform, this caused that simulation to fail and thus the test case failed. I've now updated the test to enhance it in the way in which simulates the cross platform image generation. That test and the rest of them in tier1, tier2 and tier3 all pass with this change.

mlchung commented on test/jdk/tools/jlink/plugins/CDSPluginTest.java:
```diff
@@ -87,8 +90,24 @@ public static void main(String[] args) throws Throwable {
+           // separate --module-path will force the JLink task to read the ModuleTarget from
+           // the java.base module-info.class to identify the target platform for the image
+           // being generated.
+           Path jdkRoot = Path.of(jlinkPath).getParent().getParent();
```
Or simply get from `test.jdk` system property:

```suggestion
          Path jdkRoot = Path.of(System.getProperty("test.jdk"));
```

jaikiran replied:
Hello Mandy, the `JDKToolFinder.getJDKTool("jlink")` uses the `test.jdk` to try and find the tool and then falls back to another system property, if the tool wasn't found. That's why I had used this `getParent()` technique to locate the JDK root. But I think using just the `test.jdk` system property should be fine too, since as far as I can see in the test definition of this test, there isn't anything specific that would require the test JDK to be present in any different place than the `test.jdk` system property. I've updated the PR to use this suggested approach.

mlchung reviewed (approved):
Looks good.  Minor comment in `CDSPluginTest.java`.

AlanBateman commented on src/jdk.jlink/share/classes/jdk/tools/jlink/internal/Platform.java:
```diff
@@ -24,12 +24,31 @@
  */
-public record Platform(OperatingSystem os, Architecture arch) {
+public record Platform(OperatingSystem os, Architecture arch, ByteOrder endianness) {
+    private static final Properties KNOWN_ENDIANNESS;
```
I assume this should be renamed to something like PLATFORM_PROPERTIES as the properties are not specific to endianness now.

jaikiran replied:
Hello Alan, I've updated the PR with this and the other test related suggestions.

AlanBateman commented on test/jdk/tools/jlink/plugins/CDSPluginTest.java:
```diff
@@ -102,4 +121,20 @@ public static void main(String[] args) throws Throwable {
        }
     }
+
+    private static void copyModuleFiles(Path jdkRoot, Path targetDir, String[] modules)
```
This method copies "packaged modules" so I think should be renamed to copyPackagedModules.

AlanBateman commented on test/jdk/tools/jlink/plugins/CDSPluginTest.java:
```diff
@@ -102,4 +121,20 @@ public static void main(String[] args) throws Throwable {
+            Files.copy(moduleFile, copy);
+            if (!Files.exists(copy)) {
+                throw new AssertionError("Could not copy '" + module
+                        + "' module file to directory: " + targetDir);
```
Files.copy throws if it fails so no need for the Files.exists check here.

AlanBateman commented on test/jdk/tools/jlink/plugins/CDSPluginTest.java:
```diff
@@ -87,8 +90,24 @@ public static void main(String[] args) throws Throwable {
+           // which we then use as a --module-path during jlink image generation. using such a
+           // separate --module-path will force the JLink task to read the ModuleTarget from
+           // the java.base module-info.class to identify the target platform for the image
+           // being generated.
```
I'm a bit uncomfortable with change the test for the CDS plugin as part of this change but can live with it.

jaikiran replied:
Early on during this PR, we decided and implemented to read the `ModuleTarget` only if the java.base module's file path doesn't match that of the current platform's file path. If we remove that check and instead always read the ModuleTarget of java.base, in the JLinkTask, we won't need any changes to this test, since the JLinkTask will right find the correct target platform from the module-info.class. Furthermore, rest of the cross platform image generation would work fine too.

Do you think we should remove that file path check? I am guessing the reason why the file path checks were suggested was to reduce any chances of regression with these current changes?

AlanBateman replied:
My preference would be to drop this as it insane to run with -Dos.name=window set on Linux.

RogerRiggs replied:
I'm working on an update to jlink to use jdk.internal.misc.OperatingSystem instead of os.name (PR #12931), setting os.name on the command line will have no effect. (The test will be updated or removed).

AlanBateman reviewed (approved):
The jlink changes look okay and it improves the robustness when cross linking.

My preference would be to not change the CDSPluginTest.

jaikiran commented:
Keep open.

There's work going on in the jlink and other areas to use the newer APIs for OS name and architecture detection. Specifically for jlink there's a draft PR https://github.com/openjdk/jdk/pull/13585 which along with other changes also updates the `CDSPluginTest` to stop relying on using os.name=windows on a Linux system. I plan to refresh this current PR once those changes are merged.

TheRealMDoerr commented:
We will probably have to remove the mapping here if we treat ppc64le as a new architecture:
https://github.com/openjdk/jdk/blob/cc5c9b5da2de4229c0244169bcbd6496f68db5ab/src/jdk.jlink/share/classes/jdk/tools/jlink/internal/Platform.java#L55

RogerRiggs commented:
It seems that the endian-ness is intrinsic to the architecture, as it support for 64-bit addresses. Its not just an ancillary attribute.
This PR makes the case that identifiers for os and architecture need to cover the cross-platform uses, not just the current runtime.
Including the endianness in the architecture names/enums seems like a simple and consistent way to cover the case.
The class file attribute ModuleTarget and the jdk.internal.util.Architecture namespace should be consistent or the same. 
I'd be happy to propose an update to the jdk.internal.util.Architecture to expand the namespace and provide methods for endianness and address size.

RogerRiggs commented:
See draft PR https://github.com/openjdk/jdk/pull/14063 for a proposal

jaikiran commented:
> See draft PR #14063 for a proposal

Hello Roger, I had a look at that draft PR.

The introduction of `Architecture.byteOrder()` and `Architecture.lookupByName(archName)` in that PR, I believe is going to help us here. I could then read the `ModuleTarget` attribute and use the `Architecture.lookupByName()` to get hold of the target architecture and thus its endianness through `Architecture.byteOrder()`. That would then mean we no longer need the `target.properties` file that was previously proposed in this PR.

I haven't had a chance to refresh this PR to resolve the merge conflicts. I hope to get to it soon.

RogerRiggs commented:
PR#[14063](https://github.com/openjdk/jdk/pull/14063) to add byte order and address size to Architecture has been updated but not integrated.
It has no immediate use except perhaps this PR.
So I think it can/should wait for integration until after the fork @ RPD1.

jaikiran commented:
This PR has been updated to address the merge conflicts. Additional changes are expected and will be done once https://github.com/openjdk/jdk/pull/14063 is integrated.

jaikiran commented:
This PR is now ready for further reviews. I've updated it to now completely rely on the newly introduced `jdk.internal.util.Architecture` APIs instead of the jlink code relying on a `target.properties` for mapping architectures to their endianness.
Existing tests in `test/jdk/tools/jlink` continue to pass with these latest changes. I've introduced a new jtreg test `JLinkEndianTest` to verify one of the possible `--endian` option usage. We can't automate the cross-platform testing of this command, but at least this test should verify that it won't allow creation of images with a `--endian` which doesn't match the target platform's endianness.

tier testing is currently in progress.

RogerRiggs commented on src/jdk.jlink/share/classes/jdk/tools/jlink/internal/JlinkTask.java:
```diff
@@ -802,6 +828,64 @@ private static class ImageHelper implements ImageProvider {
+                // module's ModuleTarget attribute
+                String targetPlatformVal = readJavaBaseTargetPlatform(cf);
+                try {
+                    return Platform.parsePlatform(targetPlatformVal);
```
You may want to modify `Platform.parsePlatform(s)` to use `Architecture.lookupByName(s)` instead of doing the alias mapping itself.  Platform.java:53-56.

jaikiran replied:
That's a good point. I've updated the PR with your suggested implementation.

mlchung commented on src/jdk.jlink/share/classes/jdk/tools/jlink/internal/Jlink.java:
```diff
@@ -164,12 +165,13 @@ public JlinkConfiguration(Path output,
             this.output = output;
             this.modules = Objects.requireNonNull(modules);
-            this.endian = Objects.requireNonNull(endian);
+            this.endian = endian;
```
`JlinkConfiguration` does not need to know the endianness.   Only the image builder needs it.   Probably good to take the `endian` parameter out from `JlinkConfiguration` and instead pass the value of `--endian` to `JlinkTask::createImageProvider` like other jlink options.

jaikiran replied:
Done. The updated PR no longer has the endian value in the `JlinkConfiguration`

mlchung commented on src/jdk.jlink/share/classes/jdk/tools/jlink/internal/JlinkTask.java:
```diff
@@ -802,6 +828,64 @@ private static class ImageHelper implements ImageProvider {
+
+        // returns true if the current platform's "jmods" directory is the parent of the
+        // passed javaBasePath
+        private static boolean isJavaBaseFromCurrentPlatform(Path javaBasePath) throws IOException {
```
This method checks if we are jlinking the JMOD files from the default module path.   This is not exactly checking if `java.base` matches the current platform since there are other possible setups too.    Suggest to rename the method name to `isJavaBaseFromDefaultModulePath`.   So the comments should also be updated to reflect that.

jaikiran replied:
Renamed as suggested and updated the comments as well in the updated version of the PR.

mlchung commented on src/jdk.jlink/share/classes/jdk/tools/jlink/internal/JlinkTask.java:
```diff
@@ -802,6 +828,64 @@ private static class ImageHelper implements ImageProvider {
+        // returns true if the current platform's "jmods" directory is the parent of the
+        // passed javaBasePath
+        private static boolean isJavaBaseFromCurrentPlatform(Path javaBasePath) throws IOException {
+            Path currentPlatformJmods = getDefaultModulePath();
```
```suggestion
            Path defaultModulePath = getDefaultModulePath();
```

jaikiran replied:
Done

mlchung commented on src/jdk.jlink/share/classes/jdk/tools/jlink/internal/JlinkTask.java:
```diff
@@ -408,7 +409,7 @@ private void createImage(JlinkConfiguration config) throws Exception {
 
         // First create the image provider
-        ImageProvider imageProvider = createImageProvider(config,
+        ImageHelper imageProvider = createImageProvider(config,
```
formatting nit: line 413-415 should be adjusted to align the parameters.

mlchung commented on src/jdk.jlink/share/classes/jdk/tools/jlink/internal/JlinkTask.java:
```diff
@@ -563,7 +565,8 @@ private static ImageProvider createImageProvider(JlinkConfiguration config,
         Map<String, Path> mods = cf.modules().stream()
             .collect(Collectors.toMap(ResolvedModule::name, JlinkTask::toPathLocation));
-        return new ImageHelper(cf, mods, config.getByteOrder(), retainModulesPath, ignoreSigning);
+        return new ImageHelper(cf, mods, config.getByteOrder(), retainModulesPath, ignoreSigning,
```
Also we can move the logic of determining the endianess of the target platform out of `ImageHelper`.   Evaluate the target platform and endianness here and pass to `ImageHelper`.

```suggestion
        // endian is the value specified via --endian and a new parameter to createImageProvider 
        Platform targetPlatform = targetPlatform(cf, modsPaths);
        ByteOrder targetOrder = endian != null ? endian : targetPlatform.arch().byteOrder();
        return new ImageHelper(cf, mods, targetOrder, retainModulesPath, ignoreSigning,
```

jaikiran replied:
I've updated the PR to use this suggestion. With this suggested change, the code is much more cleaner and I now just pass the `targetPlatform` to the `ImageHelper` constructor.

mlchung reviewed (commented):
Looks okay with a suggestion to determine the endianness outside `ImageHelper`.

jaikiran commented:
Thank you Roger and Mandy for the latest suggestions. I've updated the PR to incorporate those.

RogerRiggs reviewed (approved):
LGTM

mlchung commented on src/jdk.jlink/share/classes/jdk/tools/jlink/internal/JlinkTask.java:
```diff
@@ -408,16 +407,18 @@ private void createImage(JlinkConfiguration config) throws Exception {
-                                                          options.bindServices,
-                                                          options.verbose,
-                                                          log);
+        ImageHelper imageProvider = createImageProvider(config,
```
With the updated change, `createImageProvider` can continue to return `ImageProvider`.   This change can be reverted.

mlchung reviewed (approved):
Thanks for the update.  Looks much cleaner.

AlanBateman commented:
Overall I think this looks quite good, just isJavaBaseFromDefaultModulePath needs fixing.

jaikiran commented on src/jdk.jlink/share/classes/jdk/tools/jlink/internal/JlinkTask.java:
```diff
@@ -609,6 +626,63 @@ public Set<ModuleReference> findAll() {
+            // module path
+            return false;
+        }
+        return Files.isSameFile(javaBasePath, javaBaseInDefaultPath);
```
Hello Alan,

> Files.isSameFile checks the file contents, this doesn't look right here.

I was always under the impression that `Files.isSameFile()` only checks file paths and attributes for testing whether they are same. The API doc of `Files.isSameFile()` too don't mention about contents and looking at the implementation in `UnixFileSystemProvider` and `ZipFileSystemProvider` (for example), then don't seem to check the contents either. There's however a mention in the API doc of `Files.isSameFile()` which says:

> and depending on the implementation, may require to open or access both files.

So I'm guessing that, that's what allows `FileSystemProvider` implementations to actually check the content of the files.

I'll change this to use `toRealPath()` as you noted.

> Also, just to say that !Files.exists should be Files.notExists.

I've overlooked Files.notExists. I'll update this as well.

jaikiran replied:
Done. The PR has been updated with these changes.

AlanBateman replied:
Apologies, I somehow had Files.mismatch in my head, which does compare contents. If the files exists, Files.isSameFile is semantically equivalent to checking if the real path to both is equal, so what you had on that file was correct.

jaikiran replied:
Hello Alan,

> If the files exists, Files.isSameFile is semantically equivalent to checking if the real path to both is equal, so what you had on that file was correct.

I've updated the PR to use `Files.isSameFile()` like previously.

jaikiran commented:
Thank you everyone for the multiple back and forth reviews on this one.

jaikiran commented:
/integrate
